Iterative serialization of nodes

z1mvader opened the pull request:
<!-- Please describe your changes on the following line: -->
This serializes the nodes in an iterative way and now it doesn't crash when serializing a deep tree
---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #16696 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16749)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/servoparser/html.rs
 * @kichjang: components/script/dom/servoparser/html.rs

jdm commented:
@nox review ping!

nox commented on components/script/dom/servoparser/html.rs:
```diff
@@ -136,58 +141,57 @@ impl<'a> Serialize for &'a Node {
+        if stack.is_empty() {
+            Ok(())
+        } else {
+            Err(io::Error::new(io::ErrorKind::Other, "Stack is not empty!!!!"))
```
What's that for? If this isn't empty, this is a logic error and should probably panic instead.

z1mvader replied:
✔

nox commented on components/script/dom/servoparser/html.rs:
```diff
@@ -118,13 +118,18 @@ unsafe impl JSTraceable for HtmlTokenizer<TreeBuilder<JS<Node>, Sink>> {
+            roots.next().unwrap();
+        }
+        'traversal: for root in roots {
+            let node = root.r();
```
I don't think you need this here, just do `for node in roots`.

z1mvader replied:
✔

nox commented on components/script/dom/servoparser/html.rs:
```diff
@@ -118,13 +118,18 @@ unsafe impl JSTraceable for HtmlTokenizer<TreeBuilder<JS<Node>, Sink>> {
+        }
+        'traversal: for root in roots {
+            let node = root.r();
+            match  node.type_id() {
```
Nit: there is one space too many here, after `match`.

z1mvader replied:
✔

nox reviewed (changes requested):
This looks correct, but the commits need to be squashed together, and I would appreciate if you would write a proper explanation in the commit, specifically how `stack` works, and how the children count in its items is handled.

Commit:
Now a node is serialized in an iterative way. For this, we iterate
over the tree generated by the node in preorder using a vector as a
stack. When we call the serializer's method start_elem, we add a
tuple with the node's name and the total number of children. We
substract one from the number of children of the last element of
the vector when we serialize a node or when we call the serializer's
method end_elem. The purpose of this stack is to take into account
which nodes need to be ended by the serializer and when they need to
be ended. If the last element of the vector has no children left,
it is removed from the vector and ended by the serializer using
end_elem. Another way to think about this, is that the last element
of the vector always contains the name of the parent of the current
node during the iteration over the tree.

z1mvader commented:
OK @nox I believe it's ready

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4043)

nox commented:
```
  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:
  │ FAIL [expected PASS] Template contents should be serialized instead of template element if serializing template element. Test serializing whole document
  │   → assert_equals: template element is serialized incorrectly expected "<html><head><title>Test Document</title></head><body><template><div id=\"div1\">some text</div></template></body></html>" but got "<html><head><title>Test Document</title></head><body><template></template></body></html>"
  │ 
  │ @http://web-platform.test:8000/html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:62:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:52:1

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:
  │ FAIL [expected PASS] Template contents should be serialized instead of template element if serializing template element. Test nested template
  │   → assert_equals: template element is serialized incorrectly expected "<template><template><div id=\"div1\">some text</div></template></template>" but got "<template></template>"
  │ 
  │ @http://web-platform.test:8000/html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:44:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:32:1

  ▶ Unexpected subtest result in /html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:
  │ FAIL [expected PASS] Template contents should be serialized instead of template element if serializing template element
  │   → assert_equals: template element is serialized incorrectly expected "<template><div id=\"div1\">some text</div></template>" but got "<template></template>"
  │ 
  │ @http://web-platform.test:8000/html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:25:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/html/semantics/scripting-1/the-template-element/serializing-html-templates/outerhtml.html:16:1
```

You broke serialization of templates.

z1mvader commented:
Ok the problem is that I didn't include this part: https://github.com/servo/servo/blob/master/components/script/dom/servoparser/html.rs#L141

The thing is, We are traversing the tree in preorder and then we don't know which nodes are the children of the current node. should I try to downcast to HTMLTemplateElement and upcast to Node every node?

nox commented:
I would make `stack` a `Vec<(Root<Node>, usize)>`, and store in that the `DocumentFragment` associated with the template, instead of the template itself. Then when it's time to print the end tag of these nodes, write `</template>` if the node is a `DocumentFragment`.

nox commented:
Oh never mind, I see what the problem is. You probably want to make a new iterator, somehow.

z1mvader commented:
could you tell me what do you have in mind?

nox commented:
You need a new iterator, that does exactly the same traversal as the one returned by `traverse_preorder`, but which traverse the template's contents instead of its children.
